File: flipdot/queue.py
"""
Simple playlist-based content queue.

The server sends a complete playlist, and the client just plays it in order.
No priorities, no interruptions, no complex state management.
"""

from __future__ import annotations

import logging
import time
from collections import deque
from typing import TYPE_CHECKING

from flipdot.models import Content, Frame

if TYPE_CHECKING:
    from flipdot.config import DriverLimits

logger = logging.getLogger(__name__)


class ContentState:
    """State for currently playing content."""

    def __init__(self, content: Content):
        self.content: Content = content
        self.frame_index: int = 0
        self.loop_count: int = 0
        self.frame_start_time: float = time.time()

    @property
    def current_frame(self) -> Frame:
        """Get the current frame."""
        return self.content.frames[self.frame_index]

    @property
    def is_complete(self) -> bool:
        """Check if this content has finished playing."""
        playback = self.content.playback

        # If we're on the last frame
        if self.frame_index >= len(self.content.frames) - 1:
            # Not looping
            if not playback.loop:
                # If the frame has a duration, check if it has elapsed
                current_frame = self.current_frame
                if current_frame.duration_ms:
                    elapsed_ms = (time.time() - self.frame_start_time) * 1000
                    if elapsed_ms >= current_frame.duration_ms:
                        return True
                    return False
                # If duration is None/0, display indefinitely (not complete)
                return False

            # Looping with a count limit
            if (
                playback.loop_count is not None
                and self.loop_count >= playback.loop_count
            ):
                return True

        return False

    def advance_frame(self) -> bool:
        """
        Advance to the next frame if enough time has elapsed.

        Returns:
            True if frame was advanced, False otherwise
        """
        current_frame = self.current_frame
        duration = current_frame.duration_ms

        # If duration is None or 0, display indefinitely
        if not duration:
            return False

        elapsed_ms = (time.time() - self.frame_start_time) * 1000

        if elapsed_ms >= duration:
            # Move to next frame
            self.frame_index += 1

            # Check if we need to loop
            if self.frame_index >= len(self.content.frames):
                if self.content.playback.loop:
                    self.frame_index = 0
                    self.loop_count += 1
                    # Reset timer for the loop
                    self.frame_start_time = time.time()
                else:
                    # Don't advance past last frame
                    # Don't reset timer - let is_complete detect the end
                    self.frame_index = len(self.content.frames) - 1
            else:
                # Moving to a new frame, reset timer
                self.frame_start_time = time.time()

            return True

        return False


class ContentQueue:
    """
    Simple FIFO playlist for managing display content.

    Server sends complete playlist. Client just plays it in order.
    No merge logic, no priorities, no interruptions.
    """

    def __init__(self, limits: DriverLimits | None = None):
        """
        Initialize the content queue.

        Args:
            limits: Driver limits configuration (uses DEFAULT_LIMITS if None)
        """
        from flipdot.config import DEFAULT_LIMITS

        self.limits: DriverLimits = limits if limits is not None else DEFAULT_LIMITS
        self.current: ContentState | None = None
        self.queue: deque[ContentState] = deque()

    def set_playlist(self, playlist: list[Content]) -> None:
        """
        Replace entire queue with new playlist from server.

        First item becomes current (preserving frame timing if same ID).
        Rest go into queue in order.

        Args:
            playlist: Complete ordered playlist from server
        """
        if not playlist:
            logger.info("Received empty playlist, clearing")
            self.clear()
            return

        # Check if first item is same as current (preserve timing)
        new_current = playlist[0]
        old_current_state = self.current

        if (
            old_current_state
            and old_current_state.content.content_id == new_current.content_id
        ):
            logger.debug(f"Preserving playback state for {new_current.content_id}")
            # Keep existing state for smooth continuation
            new_current_state = old_current_state
            # But update the content object (in case frames changed)
            new_current_state.content = new_current
            # Validate frame index is still valid
            if new_current_state.frame_index >= len(new_current.frames):
                new_current_state.frame_index = 0
                new_current_state.loop_count = 0
                new_current_state.frame_start_time = time.time()
        else:
            # New content, start fresh
            logger.info(f"Starting new content: {new_current.content_id}")
            new_current_state = ContentState(new_current)

        self.current = new_current_state

        # Replace queue with rest of playlist
        self.queue.clear()
        for content in playlist[1:]:
            self.queue.append(ContentState(content))

        logger.info(
            f"Playlist updated: current={new_current.content_id}, "
            f"queue={len(self.queue)} items"
        )

    def update(self) -> Frame | None:
        """
        Update the queue state and return the current frame to display.

        This should be called in the main loop. It handles:
        - Advancing frames based on timing
        - Moving to next content when current completes

        Returns:
            The current frame to display, or None if nothing to display
        """
        if self.current is None:
            return None

        # Try to advance the frame
        if self.current.advance_frame():
            logger.debug(
                f"Advanced to frame {self.current.frame_index} "
                f"of {self.current.content.content_id}"
            )

        # Check if current content is complete
        if self.current.is_complete:
            logger.info(f"Content {self.current.content.content_id} completed")

            # Move to next in queue
            if self.queue:
                self.current = self.queue.popleft()
                logger.info(
                    f"Started next content: {self.current.content.content_id} "
                    f"({len(self.queue)} remaining in queue)"
                )
            else:
                # Nothing left to display
                self.current = None
                logger.info("Playlist complete, nothing to display")
                return None

        return self.current.current_frame if self.current else None

    def clear(self) -> None:
        """Clear all content from the queue."""
        logger.info("Clearing content queue")
        self.current = None
        self.queue.clear()

    def has_content(self) -> bool:
        """Check if there's any content to display."""
        return self.current is not None

    def get_current_content_id(self) -> str | None:
        """Get the ID of the currently playing content."""
        return self.current.content.content_id if self.current else None

File: driver-rs/src/queue.rs
use crate::models::{Content, Frame};
use std::collections::VecDeque;
use std::time::{Duration, Instant};

/// Per-content playback tracking. Time is injected via `now: Instant` so
/// tests stay deterministic without hitting the system clock.
pub struct ContentState {
    content: Content,
    frame_index: usize,
    loop_count: u32,
    frame_start_time: Instant,
}

impl ContentState {
    pub fn new(content: Content, now: Instant) -> Self {
        Self {
            content,
            frame_index: 0,
            loop_count: 0,
            frame_start_time: now,
        }
    }

    pub fn frame_index(&self) -> usize {
        self.frame_index
    }
    pub fn loop_count(&self) -> u32 {
        self.loop_count
    }
    pub fn content_id(&self) -> &str {
        &self.content.content_id
    }
    pub fn current_frame(&self) -> &Frame {
        &self.content.frames[self.frame_index]
    }

    pub fn advance_frame(&mut self, now: Instant) -> bool {
        let Some(duration_ms) = self.current_frame().duration_ms else {
            return false;
        };
        if duration_ms == 0 {
            return false;
        }
        let elapsed = now.saturating_duration_since(self.frame_start_time);
        if elapsed < Duration::from_millis(duration_ms as u64) {
            return false;
        }

        self.frame_index += 1;
        let n = self.content.frames.len();
        if self.frame_index >= n {
            if self.content.playback.looping {
                self.frame_index = 0;
                self.loop_count += 1;
                self.frame_start_time = now;
            } else {
                // Clamp to last frame; is_complete() detects end-of-content.
                self.frame_index = n - 1;
            }
        } else {
            self.frame_start_time = now;
        }
        true
    }

    pub fn is_complete(&self, now: Instant) -> bool {
        let n = self.content.frames.len();
        let playback = &self.content.playback;
        if self.frame_index >= n - 1 {
            if !playback.looping {
                if let Some(d) = self.current_frame().duration_ms {
                    if d == 0 {
                        return false;
                    }
                    let elapsed = now.saturating_duration_since(self.frame_start_time);
                    return elapsed >= Duration::from_millis(d as u64);
                }
                return false;
            }
            if let Some(limit) = playback.loop_count {
                if self.loop_count >= limit {
                    return true;
                }
            }
        }
        false
    }

    /// Swap the underlying Content (when a new playlist shares content_id at
    /// position 0) while preserving playback timing. Resets if frame index
    /// is no longer valid for the new frame count.
    fn replace_content(&mut self, new_content: Content, now: Instant) {
        self.content = new_content;
        if self.frame_index >= self.content.frames.len() {
            self.frame_index = 0;
            self.loop_count = 0;
            self.frame_start_time = now;
        }
    }
}

pub struct ContentQueue {
    current: Option<ContentState>,
    queue: VecDeque<ContentState>,
}

impl Default for ContentQueue {
    fn default() -> Self {
        Self::new()
    }
}

impl ContentQueue {
    pub fn new() -> Self {
        Self {
            current: None,
            queue: VecDeque::new(),
        }
    }

    pub fn set_playlist(&mut self, playlist: Vec<Content>, now: Instant) {
        if playlist.is_empty() {
            self.clear();
            return;
        }
        let mut iter = playlist.into_iter();
        let first = iter.next().expect("non-empty above");

        // Preserve state when first content_id matches current.
        match self.current.take() {
            Some(mut s) if s.content_id() == first.content_id => {
                s.replace_content(first, now);
                self.current = Some(s);
            }
            _ => {
                self.current = Some(ContentState::new(first, now));
            }
        }

        self.queue.clear();
        for c in iter {
            self.queue.push_back(ContentState::new(c, now));
        }
    }

    /// Advance the queue and return the frame to render (or None).
    pub fn update(&mut self, now: Instant) -> Option<Frame> {
        let current = self.current.as_mut()?;
        let _ = current.advance_frame(now);
        if current.is_complete(now) {
            self.current = self.queue.pop_front();
        }
        self.current.as_ref().map(|s| s.current_frame().clone())
    }

    pub fn clear(&mut self) {
        self.current = None;
        self.queue.clear();
    }

    pub fn current_content_id(&self) -> Option<&str> {
        self.current.as_ref().map(|s| s.content_id())
    }

    pub fn current_frame_index(&self) -> Option<usize> {
        self.current.as_ref().map(|s| s.frame_index())
    }

    pub fn has_content(&self) -> bool {
        self.current.is_some()
    }
}
